### ISSUE TITLE
cross platform ssh session not working

### ISSUE THREAD

**Issue by hgkamath** [wez/wezterm#507]
## Describe the bug
Cross platform ssh not working
Attempt to do a wezterm initiated ssh session from virtualbox/fedora-33-linux-VM to host/Win10 failing

The same version of wezterm is installed on both linux and windows.
I would argue that small differences in the wezterm version should not matter, and both linux and windows should be able to interpret a terminal-data-stream of either that of linux or that of windows.

ssh-ing from inside the terminal, which does work, should not be too different from a direct wezterm initiated ssh session.

## Environment (please complete the following information):

 - OS: [e.g. Linux X11, Linux Wayland, macOS, Windows]
 - Version: please run `wezterm -V` and include its output here
``` 
wezterm -V
wezterm 20210203-095643-70a364eb 
```

## To Reproduce

Steps to reproduce the behavior.
- ensure there is an ssh domain in the config
- give command ```wezterm connect host-pc```
- enter password in wezterm window

## Configuration
```lua
  ssh_domains = {
    {
      name="host-pc",
      remote_address="10.20.30.2",
      username="gana",
    },
  },
```
Be sure to include the relevant section(s) of your `wezterm.lua` configuration file.

## Expected behavior
Wezterm ssh from windows to linux should work
Wezterm ssh from linux to windows should work

## server side logs
```
C:\Users\gana>E:\scoopg\apps\wezterm\current\wezterm-mux-server
 2021-03-02T05:10:42.809Z INFO  wezterm_mux_server_impl::local > setting up C:\Users\gana\.local/share/wezterm\sock
 2021-03-02T05:11:29.279Z ERROR wezterm_mux_server_impl::local > decoding a PDU: reading PDU length: An existing connection was forcibly closed by the remote host. (os error 10054)
 2021-03-02T05:11:29.343Z ERROR wezterm_mux_server_impl::local > decoding a PDU: reading PDU length: EOF while reading leb128 encoded value
```
## client side logs
```
$ wezterm connect host-pc
 2021-03-02T04:51:23.591Z WARN  mux::ssh > while attempting agent auth: [Session(-34)] no identities found in the ssh agent
 2021-03-02T04:51:25.420Z INFO  wezterm_gui::gui::termwindow > OpenGL initialized! llvmpipe (LLVM 11.0.0, 256 bits) OpenGL ES 3.2 Mesa 20.3.4 is_context_loss_possible=false wezterm version: 20210203-095643-70a364eb
 2021-03-02T04:51:30.153Z ERROR wezterm_client::client       > going to run wezterm cli proxy
 2021-03-02T04:51:39.023Z ERROR wezterm_client::client       > ssh stderr:  2021-03-02T04:51:37.753Z ERROR wezterm_client::client > While connecting to C:\Users\gana\.local/share/wezterm\sock: No connection could be made because the target machine actively refused it. (os error 10061).  Will try spawning the server.

 2021-03-02T04:51:39.085Z ERROR wezterm_client::client       > Error while decoding response pdu: decoding a PDU: reading PDU length: Timed out waiting on socket
 2021-03-02T04:51:39.086Z ERROR wezterm_gui                  > Please install the same version of wezterm on both the client and server! The server reported error 'Error while decoding response pdu: decoding a PDU: reading PDU length: Timed out waiting on socket' while being asked for its version.  This likely means that the server is older than the client.
; terminating
 2021-03-02T04:51:39.087Z ERROR mux::connui                  > while running ConnectionUI loop: recv_timeout: channel is empty and disconnected

```

**Comment by wez** [comment]
I think the messaging about the version mismatch is a mis-attributed error.

This part of the client side logs:

```
2021-03-02T04:51:30.153Z ERROR wezterm_client::client       > going to run wezterm cli proxy
 2021-03-02T04:51:39.023Z ERROR wezterm_client::client       > ssh stderr:  2021-03-02T04:51:37.753Z ERROR
 wezterm_client::client > While connecting to C:\Users\gana\.local/share/wezterm\sock: No connection could be made
 because the target machine actively refused it. (os error 10061).  Will try spawning the server.
```

is normally where the client would launch `wezterm-mux-server` on your behalf.  If the server was already running then that error message indicates that the client doesn't have permission to connect to the socket.

I wonder if the issue might be that `wezterm-mux-server` was started as an administrator and the client doesn't have access to it?

I'd suggest terminating the manually launched `wezterm-mux-server` and trying connecting again and see if the client error message is different.

**Comment by hgkamath** [comment]
I started wezterm-muxserver as the client logs seems to suggest I should run something  on the host
The ssh fails even when no wezterm or wezterm-mux-server process is started on the host. 
Below, is a fresh log capture on client with no server running

**Comment by hgkamath** [comment]
wezterm is installed via scoop
it seems to find the wezterm socket
scoop has a wezterm shim in Scoop\shims
I tried creating an additional weaterm-mux-server shim, no difference.
When logging in via ssh the shims are in the $env:PATH. Note that my ssh login is setup with powershell as the default shell.
I think the linux wezterm client can invoke the wezterm-mux-server regardless of whether  the wezterm-mux-server shim is present, (with only the wezterm shim)
Don't know how it does that.

**Comment by wez** [comment]
Could you give the latest nightly build a try?  I think the main blocker was 9745a24 but there have been a number of improvements around ssh that I think will help too!

**Comment by hgkamath** [comment]
am checking ...

**Comment by hgkamath** [comment]
!! some progress but no success

Then got the nightly from 
Overwrote the windows scoop app directory, Updated the VM/linux fc33 rpm

```
[gana@antares ]$ wezterm -V
wezterm 20210314-114017-04b7cedd-116-g1af47804

[gana@antares ]$ wezterm connect host-pc  
 2021-03-29T15:28:37.312Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 11.0.0, 256 bits) OpenGL ES 3.2 Mesa 20.3.5 is_context_loss_possible=false wezterm version: 20210314-114017-04b7cedd-116-g1af47804
 2021-03-29T15:28:40.476Z ERROR wezterm_client::client  > going to run wezterm cli proxy
 2021-03-29T15:28:42.524Z ERROR wezterm_client::client  > ssh stderr:  2021-03-29T15:28:41.673Z ERROR wezterm_client::client > While connecting to C:\Users\gana\.local/share/wezterm\sock: No connection could be made because the target machine actively refused it. (os error 10061).  Will try spawning the server.

 2021-03-29T15:28:43.106Z ERROR wezterm_client::client  > ssh stderr:  2021-03-29T15:28:42.254Z ERROR wezterm_client::client > stdout: 
 2021-03-29T15:28:42.254Z ERROR wezterm_client::client > stderr: 

 2021-03-29T15:28:43.174Z ERROR wezterm_client::client  > ssh stderr:  2021-03-29T15:28:42.323Z INFO  wezterm_mux_server_impl::local > setting up C:\Users\gana\.local/share/wezterm\sock

 2021-03-29T15:28:43.445Z ERROR wezterm_client::client  > ssh stderr:  2021-03-29T15:28:42.594Z ERROR wezterm_mux_server_impl::local > decoding a PDU: reading PDU length: EOF while reading leb128 encoded value

 2021-03-29T15:28:43.553Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 11.0.0, 256 bits) OpenGL ES 3.2 Mesa 20.3.5 is_context_loss_possible=false wezterm version: 20210314-114017-04b7cedd-116-g1af47804
 2021-03-29T15:28:43.749Z ERROR wezterm_client::client  > got serial 111 without a corresponding promise
```

Observations 
- Both windows host, linux have the same nightly version
- Wezterm moves past the ssh login prompt
- Wezterm closes the ssh login prompt terminal window 
- Wezterm Opens another windows with a tab in it. 
- In my theme, the color changes from the regular light background to a darker shade, with an overlaid message in the top right corner "wezterm: 17s⏳since last response
- Clicking the + does not seem to open a new tab
- In windows task manager, two additional ssh process can be seen to have been spawned.  A wezterm.exe, wezterm-mux-server.exe also seem to have been spawned
- The spawned process die quickly soon after if the wezterm is Ctrl-C killed in the gnome-terminal in which it was started.

**Comment by hgkamath** [comment]
If I manually pre-start the wezterm-mux-server, in a command window (or in a wezterm terminal window with powershell) the ssh session succeed

```
C:\Users\gana>wezterm-mux-server
 2021-03-29T16:12:55.493Z INFO  wezterm_mux_server_impl::local > setting up C:\Users\gana\.local/share/wezterm\sock
 2021-03-29T16:13:07.695Z ERROR wezterm_mux_server_impl::local > decoding a PDU: reading PDU length: EOF while reading leb128 encoded value
```

**Comment by wez** [comment]
Would you mind updating both the client and server to the latest nightly and let me know how things work out for you?

**Comment by hgkamath** [comment]
Same wezterm version is installed on both host-windows-10 and guest-fedora-34
```
wezterm -V
wezterm 20210502-154244-3f7122cb
```

Without wezterm mux server prestarted
```
[gana@antares ~]$ wezterm connect host-pc
 2021-07-02T04:01:16.925Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 12.0.0, 256 bits) OpenGL ES 3.2 Mesa 21.1.3 is_context_loss_possible=false wezterm version: 20210502-154244-3f7122cb
 2021-07-02T04:01:21.721Z ERROR wezterm_client::client  > going to run wezterm cli proxy
 2021-07-02T04:01:27.315Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:01:27.960Z WARN  wezterm_client::client > While connecting to C:\Users\gana\.local/share/wezterm\sock: No connection could be made because the target machine actively refused it. (os error 10061).  Will try spawning the server.

 2021-07-02T04:01:29.911Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:01:30.556Z INFO  wezterm_mux_server_impl::local > setting up C:\Users\gana\.local/share/wezterm\sock

 2021-07-02T04:01:30.251Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:01:30.895Z ERROR wezterm_mux_server_impl::local > decoding a PDU: reading PDU length: EOF while reading leb128 encoded value

 2021-07-02T04:01:30.270Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 12.0.0, 256 bits) OpenGL ES 3.2 Mesa 21.1.3 is_context_loss_possible=false wezterm version: 20210502-154244-3f7122cb
 2021-07-02T04:01:30.641Z ERROR wezterm_client::client  > got serial 111 without a corresponding promise
```

with wezterm mux-server
```
PS C:\Users\gana> E:\scoopg\apps\wezterm\current\wezterm-mux-server.exe
 2021-07-02T04:08:07.067Z INFO  wezterm_mux_server_impl::local > setting up C:\Users\gana\.local/share/wezterm\sock
 2021-07-02T04:08:26.353Z ERROR wezterm_mux_server_impl::local > decoding a PDU: reading PDU length: EOF while reading leb128 encoded value
 2021-07-02T04:08:38.891Z ERROR wezterm_mux_server_impl::local > decoding a PDU: reading PDU length: An existing connection was forcibly closed by the remote host. (os error 10054)

[gana@antares ~]$ wezterm connect host-pc
 2021-07-02T04:08:20.961Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 12.0.0, 256 bits) OpenGL ES 3.2 Mesa 21.1.3 is_context_loss_possible=false wezterm version: 20210502-154244-3f7122cb
 2021-07-02T04:08:24.720Z ERROR wezterm_client::client  > going to run wezterm cli proxy
 2021-07-02T04:08:25.730Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 12.0.0, 256 bits) OpenGL ES 3.2 Mesa 21.1.3 is_context_loss_possible=false wezterm version: 20210502-154244-3f7122cb
```

As before ssh connection succeeds if wezterm-mux-server is pre-started

**Comment by hgkamath** [comment]
same with 
```
wezterm -V
wezterm 20210630-133255-ba42367f
```

without wezterm-mux-server
```
[gana@antares ~]$ wezterm connect capella-pc
 2021-07-02T04:37:16.950Z ERROR luahelper::serde_lua > Ignoring unknown field `font_antialias` in struct of type `Config`. Did you mean one of ...
 2021-07-02T04:37:17.086Z ERROR luahelper::serde_lua > Ignoring unknown field `font_antialias` in struct of type `Config`. Did you mean one of ...
 2021-07-02T04:37:17.103Z WARN  wezterm_ssh::config  > ssh configuration uses options that require two-phase parsing, which isn't supported
 2021-07-02T04:37:17.270Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 12.0.0, 256 bits) OpenGL ES 3.2 Mesa 21.1.3 is_context_loss_possible=false wezterm version: 20210630-133255-ba42367f
 2021-07-02T04:37:20.295Z ERROR wezterm_client::client  > going to run wezterm cli proxy
 2021-07-02T04:37:21.348Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:37:21.938Z ERROR luahelper::serde_lua > Ignoring unknown field `font_antialias` in struct of type `Config`. Did you mean one of ...
 2021-07-02T04:37:21.348Z ERROR wezterm_client::client  > ssh stderr: 
 2021-07-02T04:37:21.348Z ERROR wezterm_client::client  > ssh stderr: 

 2021-07-02T04:37:21.790Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:37:22.423Z WARN  wezterm_client::client > While connecting to C:\Users\gana\.local/share/wezterm\sock: No connection could be made because the target machine actively refused it. (os error 10061).  Will try spawning the server.

 2021-07-02T04:37:22.047Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:37:22.680Z ERROR luahelper::serde_lua > Ignoring unknown field `font_antialias` in struct of type `Config`. Did you mean one of ...
 2021-07-02T04:37:22.332Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:37:22.965Z ERROR luahelper::serde_lua > Ignoring unknown field `font_antialias` in struct of type `Config`. Did you mean one of ...
 2021-07-02T04:37:22.356Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:37:22.989Z INFO  wezterm_mux_server_impl::local > setting up C:\Users\gana\.local/share/wezterm\sock

 2021-07-02T04:37:22.595Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:37:23.228Z ERROR wezterm_mux_server_impl::local > decoding a PDU: reading PDU length: EOF while reading leb128 encoded value

 2021-07-02T04:37:22.612Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:37:23.244Z WARN  wezterm_term::terminalstate    > unhandled DecPrivateMode 9001

 2021-07-02T04:37:22.757Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 12.0.0, 256 bits) OpenGL ES 3.2 Mesa 21.1.3 is_context_loss_possible=false wezterm version: 20210630-133255-ba42367f
 2021-07-02T04:37:22.874Z ERROR wezterm_client::client  > got serial 111 without a corresponding promise
^C

```

with wezterm-mux-server
```
[gana@antares ~]$ wezterm connect capella-pc
 2021-07-02T04:38:41.201Z ERROR luahelper::serde_lua > Ignoring unknown field `font_antialias` in struct of type `Config`. Did you mean one of ...
 2021-07-02T04:38:41.224Z ERROR luahelper::serde_lua > Ignoring unknown field `font_antialias` in struct of type `Config`. Did you mean one of ...
 2021-07-02T04:38:41.235Z WARN  wezterm_ssh::config  > ssh configuration uses options that require two-phase parsing, which isn't supported
 2021-07-02T04:38:41.377Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 12.0.0, 256 bits) OpenGL ES 3.2 Mesa 21.1.3 is_context_loss_possible=false wezterm version: 20210630-133255-ba42367f
 2021-07-02T04:38:43.861Z ERROR wezterm_client::client  > going to run wezterm cli proxy
 2021-07-02T04:38:44.922Z ERROR wezterm_client::client  > ssh stderr:  2021-07-02T04:38:45.514Z ERROR luahelper::serde_lua > Ignoring unknown field `font_antialias` in struct of type `Config`. Did you mean one of 
 2021-07-02T04:38:44.922Z ERROR wezterm_client::client  > ssh stderr: swap_backspace_and_delete`, 

 2021-07-02T04:38:44.979Z INFO  wezterm_gui::termwindow > OpenGL initialized! llvmpipe (LLVM 12.0.0, 256 bits) OpenGL ES 3.2 Mesa 21.1.3 is_context_loss_possible=false wezterm version: 20210630-133255-ba42367f
[gana@antares ~]$ 

```
As before, the ssh connection succeeds if the wezterm-mux-server is pre-started

ps. the logs also indicate that in this version there is some change with 'font-alias' config I need to figure out what to do with it.

**Comment by wez** [comment]
> It looks like `wezterm imgcat` doesn't work with ssh domain
> 
> wezterm-mux-server just keeps using 100% CPU with rendering image
> 
> ![image](https://user-images.githubusercontent.com/15676035/137626253-6c0fbeed-1c40-46cd-bf7a-a405d0ebefe0.png)
> 
> If I start wezterm.exe ( without connect my.server ), and ssh to linux, then `wezterm imgcat` works, is this expected ? ![image](https://user-images.githubusercontent.com/15676035/137626351-c49260d9-9481-4e26-849f-540eb9bf2c26.png)

@xiangpeng2008  Could you file a separate issue for this?

**Comment by wez** [comment]
@hgkamath would you mind giving the nightly build a try in about 45 minutes or so?
I made some changes as part of looking into https://github.com/wez/wezterm/issues/1358 that might also help with this one

**Comment by hgkamath** [comment]
I'll try my best to get to it within a couple of days. presently computer in semi-configured state as I juggle disks around, trying to move above from raw partitions to vdisks. I'm also investigating switching from vbox to whpx/qemu.  I'll update in this comment when I set things back up.
